src/DbalPersistence.php

```diff
@@ -185,39 +186,47 @@ public function commit(
-                'aggregate_root_type' => $aggregateRootType->getContractName(),
-                'aggregate_root_id' => $aggregateRootId,
-                'aggregate_root_version' => $aggregateRootVersion,
-                'event_type' => $eventEnvelope->getEventType()->getContractName(),
-                'event_id' => $eventEnvelope->getEventId(),
-                'event' => json_encode(
-                    $this->eventSerializer->serialize(
-                        $eventEnvelope->getEventType(),
-                        $eventEnvelope->getEvent()
+        $this->connection->beginTransaction();
+
+        try {
+            foreach ($eventEnvelopes as $eventEnvelope) {
+                $metadata = $eventEnvelope->getMetadataType()
+                    ? json_encode(
+                        $this->metadataSerializer->serialize(
+                            $eventEnvelope->getMetadataType(),
+                            $eventEnvelope->getMetadata()
+                        )
                     )
-                ),
-                'event_version' => $eventEnvelope->getVersion(),
-                '`when`' => $eventEnvelope->getWhen()->format('Y-m-d H:i:s'),
-                'metadata_type' => $eventEnvelope->getMetadataType()
-                    ? $eventEnvelope->getMetadataType()->getContractName()
-                    : null,
-                'metadata' => $metadata,
-            ];
-            $this->connection->insert($this->tableName, $values);
+                    : null;
+                $values = [
+                    'commit_id' => $commitId,
+                    'utc_committed_time' => $now->format('Y-m-d H:i:s'),
+                    'aggregate_root_type' => $aggregateRootType->getContractName(),
+                    'aggregate_root_id' => $aggregateRootId,
+                    'aggregate_root_version' => $aggregateRootVersion,
+                    'event_type' => $eventEnvelope->getEventType()->getContractName(),
+                    'event_id' => $eventEnvelope->getEventId(),
+                    'event' => json_encode(
+                        $this->eventSerializer->serialize(
+                            $eventEnvelope->getEventType(),
+                            $eventEnvelope->getEvent()
+                        )
+                    ),
+                    'event_version' => $eventEnvelope->getVersion(),
+                    '`when`' => $eventEnvelope->getWhen()->format('Y-m-d H:i:s'),
+                    'metadata_type' => $eventEnvelope->getMetadataType()
+                        ? $eventEnvelope->getMetadataType()->getContractName()
+                        : null,
+                    'metadata' => $metadata,
+                ];
+                $this->connection->insert($this->tableName, $values);
+            }
+
+            $this->connection->commit();
+        } catch (\Exception $e) {
+            $this->connection->rollBack();
 
+            throw $e;
```

Comment:
Looks like we might be able to do some magic with PDO to throw an `OptimisticConcurrencyFailed` exception.

Before simply throwing `$e`, do something like:

```
if ($e instanceof \Doctrine\DBAL\DBALException) {
  if ($e->getPrevious() && in_array($e->getPrevious()->getCode(), [23000, 23050])) {
    throw new OptimisticConcurrencyFailed();
  }
}
```

MySQL & SQLite use error code 23000 and Postgresql uses 23050 for duplicate errors. This would allow us to be more specific with the type of exception we are receiving.

While we are in this part of the code I think it would be good to enhance `OptimisticConcurrencyFailed` to be more useful. I think it should have a constructor signature of `__construct($aggregateRootType, $aggregateRootId, $message = null, \Exception $previous = null)`.

The earlier call to `new OptimisticConcurrencyFailed()` would be changed to:

```
new OptimisticConcurrencyFailed(
  $aggregateRootType->getContractName(),
  $aggregateRootId,
  sprintf(
    'Expected aggregate root version %d but found %d.',
    $expectedAggregateRootVersion,
    $aggregateRootVersion
  )
)
```

The newer one would become:

```
throw new OptimisticConcurrencyFailed(
  $aggregateRootType->getContractName(),
  $aggregateRootId,
  'Duplicate event version.',
  $e
);
```

Inside `OptimisticConcurrencyFailed`'s constructor, format the message nicely to include the aggregate root type and id:

```
public function __construct($aggregateRootType, $aggregateRootId, $message = null, \Exception $previous = null)
{
  $formattedMessage = sprintf('%s (%s:%s)', $message, $aggregateRootType, $aggregateRootId);
  parent::__construct($formattedMessage, 0, $previous);
}
```

I believe `OptimisticConcurrencyFailed` lives in the main `depot-event-store` repository so you'll need to go send a PR there first and then `composer update depot/event-store` after that PR is merged. Make sure to update the InMemory implementation in that repo too as I believe it throws `OptimisticConcurrencyFailed` itself as well.